numpy/_core/src/umath/matmul.c.src

```diff
@@ -515,3 +515,141 @@ NPY_NO_EXPORT void
+            ip2 += chunk * is2;
+            n -= chunk;
+        }
+        ((@type@ *)op)[0] = (@type@)sum[0];
+        ((@type@ *)op)[1] = (@type@)sum[1];
+    }
+    else
+#endif
+    {
+        @sumtype@ sumr = 0., sumi = 0.;  /* double for stability */
+        for (npy_intp i = 0; i < n; i++, ip1 += is1, ip2 += is2) {
+            const @type@ ip1r = ((@type@ *)ip1)[0];
+            const @type@ ip1i = ((@type@ *)ip1)[1];
+            const @type@ ip2r = ((@type@ *)ip2)[0];
+            const @type@ ip2i = ((@type@ *)ip2)[1];
+            /* ip1.conj() * ip2 */
+            sumr += (double)(ip1r * ip2r + ip1i * ip2i);
+            sumi += (double)(ip1r * ip2i - ip1i * ip2r);
+        }
+        ((@type@ *)op)[0] = (@type@)sumr;
+        ((@type@ *)op)[1] = (@type@)sumi;
+    }
+}
+/**end repeat**/
+
+static void
+OBJECT_dotc(char *ip1, npy_intp is1, char *ip2, npy_intp is2, char *op, npy_intp n,
+            void *NPY_UNUSED(ignore))
+{
+    npy_intp i;
+    PyObject *result = NULL;
+
+    for (i = 0; i < n; i++, ip1 += is1, ip2 += is2) {
+        PyObject *obj1 = *(PyObject**)ip1, *obj2 = *(PyObject**)ip2;
+        if (obj1 == NULL) {
+            obj1 = Py_None;  /* raise useful error message at conjugate */
+        }
+        if (obj2 == NULL) {
+            obj2 = Py_None;  /* raise at multiply (unless obj1 handles None) */
+        }
+        PyObject *obj1_conj = PyObject_CallMethod(obj1, "conjugate", NULL);
+        if (obj1_conj == NULL) {
+            goto fail;
+        }
+        PyObject *prod = PyNumber_Multiply(obj1_conj, obj2);
+        Py_DECREF(obj1_conj);
+        if (prod == NULL) {
+            goto fail;
+        }
+        if (i == 0) {
+            result = prod;
+        }
+        else {
+            Py_SETREF(result, PyNumber_Add(result, prod));
+            Py_DECREF(prod);
+            if (result == NULL) {
+                goto fail;
+            }
+        }
+    }
+    Py_XSETREF(*((PyObject **)op), result);
+    return;
+
+  fail:
+    Py_XDECREF(result);
+    return;
+}
+
+/**begin repeat
+ *  #TYPE = FLOAT, DOUBLE, LONGDOUBLE, HALF,
+ *          UBYTE, USHORT, UINT, ULONG, ULONGLONG,
+ *          BYTE, SHORT, INT, LONG, LONGLONG, BOOL,
+ *          CFLOAT, CDOUBLE, CLONGDOUBLE, OBJECT#
+ *  #DOT = dot*15, dotc*4#
+ */
+NPY_NO_EXPORT void
+@TYPE@_vecdot(char **args, npy_intp const *dimensions, npy_intp const *steps,
+              void *NPY_UNUSED(func))
+{
+    npy_intp n_outer = dimensions[0];
+    npy_intp s0=steps[0], s1=steps[1], s2=steps[2];
+    npy_intp n_inner = dimensions[1];
+    npy_intp is1=steps[3], is2=steps[4];
+
+    for (npy_intp i = 0; i < n_outer; i++, args[0] += s0, args[1] += s1, args[2] += s2) {
+        /*
+         * TODO: use new API to select inner loop with get_loop and
+         * improve error treatment.
+         */
+        char *ip1=args[0], *ip2=args[1], *op=args[2];
+        @TYPE@_@DOT@(ip1, is1, ip2, is2, op, n_inner, NULL);
+#if @TYPE@ == OBJECT
```

Comment:
Hmmm, does this work if OBJECT is undefined, I think?  I sometimes use `#define IS_@TYPE@`

Reply:
OK, just goes to show how I don't really grok pre-processing... I just added an extra iteration item...